File: FBControlCoreTests/Tests/Unit/FBFutureContextManagerTests.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import <XCTest/XCTest.h>

#import <FBControlCore/FBControlCore.h>

@interface FBFutureContextManagerTests : XCTestCase <FBFutureContextManagerDelegate>

@property (nonatomic, strong, readwrite) dispatch_queue_t queue;
@property (nonatomic, assign, readwrite) NSUInteger prepareCalled;
@property (atomic, assign, readwrite) NSUInteger teardownCalled;
@property (nonatomic, copy, readwrite) NSNumber *contextPoolTimeout;
@property (nonatomic, assign, readwrite) BOOL failPrepare;
@property (nonatomic, assign, readwrite) BOOL resetFailPrepare; // if YES, set failPrepare = NO when prepare fails
@property (nonatomic, assign, readwrite) BOOL isContextSharable;
@property (nonatomic, strong, readwrite) id<FBControlCoreLogger> logger;

@end

@implementation FBFutureContextManagerTests

- (void)setUp
{
  self.queue = dispatch_queue_create("com.facebook.fbcontrolcore.tests.future_context", DISPATCH_QUEUE_SERIAL);
  self.logger = [FBControlCoreGlobalConfiguration.defaultLogger withName:@"manager_test"];
  self.isContextSharable = NO;
  self.contextPoolTimeout = nil;
  self.prepareCalled = 0;
  self.teardownCalled = 0;
  self.failPrepare = NO;
  self.resetFailPrepare = NO;
}

- (FBFutureContextManager<NSNumber *> *)manager
{
  return [FBFutureContextManager managerWithQueue:self.queue delegate:self logger:self.logger];
}

- (void)testSingleAquire
{
  FBFuture *future = [[self.manager
    utilizeWithPurpose:@"A Test"]
    onQueue:self.queue pop:^(id result) {
      return [FBFuture futureWithResult:@123];
    }];

  NSError *error = nil;
  id value = [future awaitWithTimeout:1 error:&error];
  XCTAssertNil(error);
  XCTAssertEqualObjects(value, @123);

  XCTAssertEqual(self.prepareCalled, 1u);
  XCTAssertEqual(self.teardownCalled, 1u);
}

- (void)testSequentialAquire
{
  FBFutureContextManager<NSNumber *> *manager = self.manager;

  FBFuture *future0 = [[manager
    utilizeWithPurpose:@"A Test"]
    onQueue:self.queue pop:^(id result) {
      return [FBFuture futureWithResult:@0];
    }];

  NSError *error = nil;
  id value = [future0 awaitWithTimeout:1 error:&error];
  XCTAssertNil(error);
  XCTAssertEqualObjects(value, @0);

  XCTAssertEqual(self.prepareCalled, 1u);
  XCTAssertEqual(self.teardownCalled, 1u);

  FBFuture *future1 = [[manager
    utilizeWithPurpose:@"A Test"]
    onQueue:self.queue pop:^(id result) {
      return [FBFuture futureWithResult:@1];
    }];
  value = [future1 awaitWithTimeout:1 error:&error];
  XCTAssertNil(error);
  XCTAssertEqualObjects(value, @1);

  XCTAssertEqual(self.prepareCalled, 2u);
  XCTAssertEqual(self.teardownCalled, 2u);
}

- (void)testSequentialAquireWithCooloff
{
  FBFutureContextManager<NSNumber *> *manager = self.manager;
  self.contextPoolTimeout = @0.2;

  FBFuture *future0 = [[manager
    utilizeWithPurpose:@"A Test"]
    onQueue:self.queue pop:^(id result) {
      return [FBFuture futureWithResult:@0];
    }];

  NSError *error = nil;
  id value = [future0 awaitWithTimeout:1 error:&error];
  XCTAssertNil(error);
  XCTAssertEqualObjects(value, @0);

  XCTAssertEqual(self.prepareCalled, 1u);
  XCTAssertEqual(self.teardownCalled, 0u);

  FBFuture *future1 = [[manager
    utilizeWithPurpose:@"A Test"]
    onQueue:self.queue pop:^(id result) {
      return [FBFuture futureWithResult:@1];
    }];
  value = [future1 awaitWithTimeout:1 error:&error];
  XCTAssertNil(error);
  XCTAssertEqualObjects(value, @1);

  XCTAssertEqual(self.prepareCalled, 1u);
  XCTAssertEqual(self.teardownCalled, 0u);

  [[FBFuture futureWithDelay:0.25 future:FBFuture.empty] await:nil];

  XCTAssertEqual(self.prepareCalled, 1u);
  XCTAssertEqual(self.teardownCalled, 1u);
}

- (void)testConcurrentAquireOnlyPreparesOnce
{
  FBFutureContextManager<NSNumber *> *manager = self.manager;
  dispatch_queue_t concurrent = dispatch_queue_create("com.facebook.fbcontrolcore.tests.future_context.concurrent", DISPATCH_QUEUE_CONCURRENT);
  FBMutableFuture *future0 = FBMutableFuture.future;
  FBMutableFuture *future1 = FBMutableFuture.future;
  FBMutableFuture *future2 = FBMutableFuture.future;

  dispatch_async(concurrent, ^{
    FBFuture *inner = [[manager
      utilizeWithPurpose:@"Test 0"]
      onQueue:self.queue pop:^(id result) {
        [self.logger log:@"Test 0 In Use"];
        return [FBFuture futureWithResult:@0];
      }];
    [future0 resolveFromFuture:inner];
  });
  dispatch_async(concurrent, ^{
    FBFuture *inner = [[manager
      utilizeWithPurpose:@"Test 1"]
      onQueue:self.queue pop:^(id result) {
        [self.logger log:@"Test 1 In Use"];
        return [FBFuture futureWithResult:@1];
      }];
    [future1 resolveFromFuture:inner];
  });
  dispatch_async(concurrent, ^{
    FBFuture *inner = [[manager
      utilizeWithPurpose:@"Test 2"]
      onQueue:self.queue pop:^(id result) {
        [self.logger log:@"Test 2 In Use"];
        return [FBFuture futureWithResult:@2];
      }];
    [future2 resolveFromFuture:inner];
  });

  NSError *error = nil;
  id value = [[FBFuture futureWithFutures:@[future0, future1, future2]] awaitWithTimeout:1 error:&error];
  XCTAssertNil(error);
  XCTAssertEqualObjects(value, (@[@0, @1, @2]));

  XCTAssertEqual(self.prepareCalled, 1u);
  XCTAssertEqual(self.teardownCalled, 1u);
}

- (void)testConcurrentAquireWithSharableResource
{
  FBFutureContextManager<NSNumber *> *manager = self.manager;
  self.isContextSharable = YES;
  dispatch_queue_t concurrent = dispatch_queue_create("com.facebook.fbcontrolcore.tests.future_context.concurrent", DISPATCH_QUEUE_CONCURRENT);
  FBMutableFuture *future0 = FBMutableFuture.future;
  FBMutableFuture *future1 = FBMutableFuture.future;
  FBMutableFuture *future2 = FBMutableFuture.future;

  dispatch_async(concurrent, ^{
    FBFuture *inner = [[manager
      utilizeWithPurpose:@"Test 0"]
      onQueue:self.queue pop:^(id result) {
        [self.logger log:@"Test 0 In Use"];
        return [FBFuture futureWithResult:@0];
      }];
    [future0 resolveFromFuture:inner];
  });
  dispatch_async(concurrent, ^{
    FBFuture *inner = [[manager
      utilizeWithPurpose:@"Test 1"]
      onQueue:self.queue pop:^(id result){
        [self.logger log:@"Test 1 In Use"];
        return [FBFuture futureWithResult:@1];
      }];
    [future1 resolveFromFuture:inner];
  });
  dispatch_async(concurrent, ^{
    FBFuture *inner = [[manager
      utilizeWithPurpose:@"Test 2"]
      onQueue:self.queue pop:^(id result) {
        [self.logger log:@"Test 2 In Use"];
        return [FBFuture futureWithResult:@2];
      }];
    [future2 resolveFromFuture:inner];
  });

  NSError *error = nil;
  id value = [[FBFuture futureWithFutures:@[future0, future1, future2]] awaitWithTimeout:1 error:&error];
  XCTAssertNil(error);
  XCTAssertEqualObjects(value, (@[@0, @1, @2]));

  XCTAssertEqual(self.prepareCalled, 1u);
  XCTAssertEqual(self.teardownCalled, 1u);
}

- (void)testFailInPrepare
{
  FBFutureContextManager<NSNumber *> *manager = self.manager;
  self.failPrepare = YES;
  FBFuture *future0 = [[manager
    utilizeWithPurpose:@"A Test"]
    onQueue:self.queue pop:^(id result) {
      return [FBFuture futureWithResult:@0];
    }];

  NSError *error = nil;
  id value = [future0 awaitWithTimeout:1 error:&error];
  XCTAssertNotNil(error);

  XCTAssertEqual(self.prepareCalled, 1u);
  XCTAssertEqual(self.teardownCalled, 0u);

  self.failPrepare = NO;
  FBFuture *future1 = [[manager
    utilizeWithPurpose:@"A Test"]
    onQueue:self.queue pop:^(id result) {
      return [FBFuture futureWithResult:@1];
    }];
  error = nil;
  value = [future1 awaitWithTimeout:1 error:&error];
  XCTAssertNil(error);
  XCTAssertEqualObjects(value, @1);

  XCTAssertEqual(self.prepareCalled, 2u);
  XCTAssertEqual(self.teardownCalled, 1u);
}

- (void)testConcurrentAquireWithOneFailInPrepare
{
  FBFutureContextManager<NSNumber *> *manager = self.manager;
  dispatch_queue_t concurrent = dispatch_queue_create("com.facebook.fbcontrolcore.tests.future_context.concurrent", DISPATCH_QUEUE_CONCURRENT);
  FBMutableFuture *future0 = FBMutableFuture.future;
  FBMutableFuture *future1 = FBMutableFuture.future;
  FBMutableFuture *future2 = FBMutableFuture.future;

  self.failPrepare = YES;
  self.resetFailPrepare = YES;
  dispatch_async(concurrent, ^{
    FBFuture *inner = [[manager
      utilizeWithPurpose:@"A Test 1"]
      onQueue:self.queue pop:^(id result) {
       return [FBFuture futureWithResult:@0];
      }];
    [future0 resolveFromFuture:inner];
  });
  dispatch_async(concurrent, ^{
    FBFuture *inner = [[manager
    utilizeWithPurpose:@"A Test 2"]
    onQueue:self.queue pop:^(id result) {
      return [FBFuture futureWithResult:@1];
    }];
    [future1 resolveFromFuture:inner];
  });
  dispatch_async(concurrent, ^{
    FBFuture *inner = [[manager
      utilizeWithPurpose:@"A Test 3"]
      onQueue:self.queue pop:^(id result) {
        return [FBFuture futureWithResult:@2];
      }];
    [future2 resolveFromFuture:inner];
  });

  NSMutableArray<NSNumber *> *values = [NSMutableArray array];
  NSMutableArray<NSError *> *errors = [NSMutableArray array];

  for (FBFuture *future in @[future0, future1, future2]) {
    NSNumber *value = nil;
    NSError *error = nil;
    value = [future awaitWithTimeout:1 error:&error];
    if (value) {
      [values addObject:value];
    }
    if (error) {
      [errors addObject:error];
    }
  }

  XCTAssertEqual(values.count, 2u);
  XCTAssertEqual(errors.count, 1u);

  XCTAssertEqual(self.prepareCalled, 2u);
  XCTAssertEqual(self.teardownCalled, 1u);
}

- (void)testImmediateAquireAndRelease
{
  FBFutureContextManager<NSNumber *> *manager = self.manager;

  NSError *error = nil;
  NSNumber *context = [manager utilizeNowWithPurpose:@"A Test" error:&error];
  XCTAssertNil(error);
  XCTAssertEqualObjects(context, @0);

  BOOL success = [manager returnNowWithPurpose:@"A Test" error:&error];
  XCTAssertNil(error);
  XCTAssertTrue(success);
}

- (NSString *)contextName
{
  return @"A Test";
}

- (FBFuture<id> *)prepare:(id<FBControlCoreLogger>)logger
{
  self.prepareCalled++;
  if (self.failPrepare) {
    if (self.resetFailPrepare) {
      self.failPrepare = NO;
    }
    return [[FBControlCoreError describe:@"Error in prepare"] failFuture];
  }
  else {
    return [FBFuture futureWithResult:@0];
  }
}

- (FBFuture<NSNull *> *)teardown:(id)context logger:(id<FBControlCoreLogger>)logger
{
  self.teardownCalled++;
  return FBFuture.empty;
}

@end

File: FBControlCoreTests/Tests/Unit/FBFutureContextManagerTests.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import XCTest

@testable import FBControlCore

final class FBFutureContextManagerTests: XCTestCase, FBFutureContextManagerDelegate {
  var queue: DispatchQueue!
  var prepareCalled: UInt = 0
  var teardownCalled: UInt = 0
  var contextPoolTimeout: NSNumber? = 0
  var failPrepare: Bool = false
  var resetFailPrepare: Bool = false
  var isContextSharable: Bool = false
  var logger: FBControlCoreLogger!

  override func setUp() {
    queue = DispatchQueue(label: "com.facebook.fbcontrolcore.tests.future_context")
    logger = FBControlCoreGlobalConfiguration.defaultLogger.withName("manager_test")
    isContextSharable = false
    contextPoolTimeout = 0
    prepareCalled = 0
    teardownCalled = 0
    failPrepare = false
    resetFailPrepare = false
  }

  var manager: FBFutureContextManager<NSNumber> {
    return FBFutureContextManager<NSNumber>(queue: queue, delegate: self, logger: logger)
  }

  func testSingleAquire() {
    let future =
      manager
      .utilize(withPurpose: "A Test")
      .onQueue(
        queue,
        pop: { result in
          return FBFuture<AnyObject>(result: NSNumber(value: 123))
        })

    let value = try? future.await(withTimeout: 1) as? NSNumber
    XCTAssertEqual(value, 123)

    XCTAssertEqual(prepareCalled, 1)
    XCTAssertEqual(teardownCalled, 1)
  }

  func testSequentialAquire() {
    let manager = self.manager

    let future0 =
      manager
      .utilize(withPurpose: "A Test")
      .onQueue(
        queue,
        pop: { result in
          return FBFuture<AnyObject>(result: NSNumber(value: 0))
        })

    var value = try? future0.await(withTimeout: 1) as? NSNumber
    XCTAssertEqual(value, 0)

    XCTAssertEqual(prepareCalled, 1)
    XCTAssertEqual(teardownCalled, 1)

    let future1 =
      manager
      .utilize(withPurpose: "A Test")
      .onQueue(
        queue,
        pop: { result in
          return FBFuture<AnyObject>(result: NSNumber(value: 1))
        })
    value = try? future1.await(withTimeout: 1) as? NSNumber
    XCTAssertEqual(value, 1)

    XCTAssertEqual(prepareCalled, 2)
    XCTAssertEqual(teardownCalled, 2)
  }

  func testConcurrentAquireOnlyPreparesOnce() {
    let manager = self.manager
    let queue = self.queue!
    let logger = self.logger!
    let concurrent = DispatchQueue(label: "com.facebook.fbcontrolcore.tests.future_context.concurrent", attributes: .concurrent)
    let future0 = FBMutableFuture<AnyObject>()
    let future1 = FBMutableFuture<AnyObject>()
    let future2 = FBMutableFuture<AnyObject>()

    let block0: @convention(block) () -> Void = {
      let inner =
        manager
        .utilize(withPurpose: "Test 0")
        .onQueue(
          queue,
          pop: { result in
            logger.log("Test 0 In Use")
            return FBFuture<AnyObject>(result: NSNumber(value: 0))
          })
      (future0 as! FBMutableFuture).resolve(from: inner)
    }
    let block1: @convention(block) () -> Void = {
      let inner =
        manager
        .utilize(withPurpose: "Test 1")
        .onQueue(
          queue,
          pop: { result in
            logger.log("Test 1 In Use")
            return FBFuture<AnyObject>(result: NSNumber(value: 1))
          })
      (future1 as! FBMutableFuture).resolve(from: inner)
    }
    let block2: @convention(block) () -> Void = {
      let inner =
        manager
        .utilize(withPurpose: "Test 2")
        .onQueue(
          queue,
          pop: { result in
            logger.log("Test 2 In Use")
            return FBFuture<AnyObject>(result: NSNumber(value: 2))
          })
      (future2 as! FBMutableFuture).resolve(from: inner)
    }

    concurrent.async(execute: block0)
    concurrent.async(execute: block1)
    concurrent.async(execute: block2)

    let value = try? FBFuture<AnyObject>.combine([future0, future1, future2]).await(withTimeout: 1) as NSArray?
    XCTAssertNotNil(value)
    XCTAssertEqual(value as? NSArray, [0, 1, 2] as NSArray)

    XCTAssertEqual(prepareCalled, 1)
    XCTAssertEqual(teardownCalled, 1)
  }

  func testConcurrentAquireWithSharableResource() {
    let manager = self.manager
    let queue = self.queue!
    let logger = self.logger!
    isContextSharable = true
    let concurrent = DispatchQueue(label: "com.facebook.fbcontrolcore.tests.future_context.concurrent", attributes: .concurrent)
    let future0 = FBMutableFuture<AnyObject>()
    let future1 = FBMutableFuture<AnyObject>()
    let future2 = FBMutableFuture<AnyObject>()

    let block0: @convention(block) () -> Void = {
      let inner =
        manager
        .utilize(withPurpose: "Test 0")
        .onQueue(
          queue,
          pop: { result in
            logger.log("Test 0 In Use")
            return FBFuture<AnyObject>(result: NSNumber(value: 0))
          })
      (future0 as! FBMutableFuture).resolve(from: inner)
    }
    let block1: @convention(block) () -> Void = {
      let inner =
        manager
        .utilize(withPurpose: "Test 1")
        .onQueue(
          queue,
          pop: { result in
            logger.log("Test 1 In Use")
            return FBFuture<AnyObject>(result: NSNumber(value: 1))
          })
      (future1 as! FBMutableFuture).resolve(from: inner)
    }
    let block2: @convention(block) () -> Void = {
      let inner =
        manager
        .utilize(withPurpose: "Test 2")
        .onQueue(
          queue,
          pop: { result in
            logger.log("Test 2 In Use")
            return FBFuture<AnyObject>(result: NSNumber(value: 2))
          })
      (future2 as! FBMutableFuture).resolve(from: inner)
    }

    concurrent.async(execute: block0)
    concurrent.async(execute: block1)
    concurrent.async(execute: block2)

    let value = try? FBFuture<AnyObject>.combine([future0, future1, future2]).await(withTimeout: 1) as NSArray?
    XCTAssertNotNil(value)
    XCTAssertEqual(value as? NSArray, [0, 1, 2] as NSArray)

    XCTAssertEqual(prepareCalled, 1)
    XCTAssertEqual(teardownCalled, 1)
  }

  func testFailInPrepare() {
    let manager = self.manager
    failPrepare = true
    let future0 =
      manager
      .utilize(withPurpose: "A Test")
      .onQueue(
        queue,
        pop: { result in
          return FBFuture<AnyObject>(result: NSNumber(value: 0))
        })

    do {
      _ = try future0.await(withTimeout: 1)
      XCTFail("Expected error")
    } catch {
      // Expected error
    }

    XCTAssertEqual(prepareCalled, 1)
    XCTAssertEqual(teardownCalled, 0)

    failPrepare = false
    let future1 =
      manager
      .utilize(withPurpose: "A Test")
      .onQueue(
        queue,
        pop: { result in
          return FBFuture<AnyObject>(result: NSNumber(value: 1))
        })
    let value = try? future1.await(withTimeout: 1) as? NSNumber
    XCTAssertEqual(value, 1)

    XCTAssertEqual(prepareCalled, 2)
    XCTAssertEqual(teardownCalled, 1)
  }

  func testConcurrentAquireWithOneFailInPrepare() {
    let manager = self.manager
    let queue = self.queue!
    let concurrent = DispatchQueue(label: "com.facebook.fbcontrolcore.tests.future_context.concurrent", attributes: .concurrent)
    let future0 = FBMutableFuture<AnyObject>()
    let future1 = FBMutableFuture<AnyObject>()
    let future2 = FBMutableFuture<AnyObject>()

    failPrepare = true
    resetFailPrepare = true

    let block0: @convention(block) () -> Void = {
      let inner =
        manager
        .utilize(withPurpose: "A Test 1")
        .onQueue(
          queue,
          pop: { result in
            return FBFuture<AnyObject>(result: NSNumber(value: 0))
          })
      (future0 as! FBMutableFuture).resolve(from: inner)
    }
    let block1: @convention(block) () -> Void = {
      let inner =
        manager
        .utilize(withPurpose: "A Test 2")
        .onQueue(
          queue,
          pop: { result in
            return FBFuture<AnyObject>(result: NSNumber(value: 1))
          })
      (future1 as! FBMutableFuture).resolve(from: inner)
    }
    let block2: @convention(block) () -> Void = {
      let inner =
        manager
        .utilize(withPurpose: "A Test 3")
        .onQueue(
          queue,
          pop: { result in
            return FBFuture<AnyObject>(result: NSNumber(value: 2))
          })
      (future2 as! FBMutableFuture).resolve(from: inner)
    }

    concurrent.async(execute: block0)
    concurrent.async(execute: block1)
    concurrent.async(execute: block2)

    var values: [NSNumber] = []
    var errors: [Error] = []

    for future in [future0, future1, future2] {
      do {
        if let value = try future.await(withTimeout: 1) as? NSNumber {
          values.append(value)
        }
      } catch {
        errors.append(error)
      }
    }

    XCTAssertEqual(values.count, 2)
    XCTAssertEqual(errors.count, 1)

    XCTAssertEqual(prepareCalled, 2)
    XCTAssertEqual(teardownCalled, 1)
  }

  func testImmediateAquireAndRelease() throws {
    let manager = self.manager

    let context = try manager.utilizeNow(withPurpose: "A Test")
    XCTAssertEqual(context as? NSNumber, 0)

    try manager.returnNow(withPurpose: "A Test")
  }

  // MARK: - FBFutureContextManagerDelegate

  var contextName: String {
    return "A Test"
  }

  func prepare(_ logger: FBControlCoreLogger) -> FBFuture<AnyObject> {
    prepareCalled += 1
    if failPrepare {
      if resetFailPrepare {
        failPrepare = false
      }
      return FBControlCoreError.describe("Error in prepare").failFuture()
    } else {
      return FBFuture<AnyObject>(result: NSNumber(value: 0))
    }
  }

  func teardown(_ context: Any, logger: FBControlCoreLogger) -> FBFuture<NSNull> {
    teardownCalled += 1
    return FBFuture<NSNull>.empty()
  }
}
